### DISCUSSION TITLE
Doing database change auditing with sea-orm

### DISCUSSION THREAD

**Discussion by daviessm** [SeaQL/sea-orm#972] [Q&A]
I've been asked to write an application with database change auditing, i.e. every UPDATE or DELETE statement also needs to write the previous version of a record to an audit table containing the record as it was before the change. Additionally I need to record the date and time of the change as well as the user who made the change (also for INSERTs) in another table.

Ideally this should happen within a database transaction so that if either the auditing of a change fails, or the actual change fails, the transaction is rolled back.

Should it be possible for me to extend sea-orm to add things like a `delete_by_user(user_id)` function to each entity which runs the above SQL?

**Comment by billy1624** [comment]
Hey @daviessm, welcome! Yes, as you said, the actual update / insert as well as the insertion of the audit log can be done inside a transaction closure: https://www.sea-ql.org/SeaORM/docs/advanced-query/transaction/#within-a-closure.

This issue could be related
- https://github.com/SeaQL/sea-orm/issues/962

**Reply by daviessm** [reply]
Thanks @billy1624. Do you think it would be simple enough for me to write a macro which creates the transaction code? I don't want to have to write that transaction block every time I save an entity to the database.

I'm thinking along the lines of adding delete_as(user_id) to [ActiveModelTrait](https://docs.rs/sea-orm/latest/sea_orm/entity/trait.ActiveModelTrait.html), or a new ActiveModelAuditTrait that provides delete_as, insert_as and update_as.

**Reply by billy1624** [reply]
Hey @daviessm, sorry for the delay. I don't think you need a macro to do that. You can create a function that take any `A: ActiveModel` then check its values column by column and see which one is being updated. Then, you can insert an audit log into the database.

Similar to what we have here, not exactly but similar
- https://github.com/SeaQL/sea-orm/blob/87f7891f175d662280ede2a234421bedac9507ca/src/entity/active_model.rs#L491-L517
